File: ThumbnailExtension/ThumbnailProvider.m
//
//  ThumbnailProvider.m
//  ThumbnailExtension
//
//  Created by Torsten Kammer on 12.11.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

#import "ThumbnailProvider.h"

#import "Gear.h"
#import "Gear+ExportAsPDF.h"

@implementation ThumbnailProvider

- (void)provideThumbnailForFileRequest:(QLFileThumbnailRequest *)request completionHandler:(void (^)(QLThumbnailReply * _Nullable, NSError * _Nullable))handler {

    NSError *error = nil;
    NSData *data = [NSData dataWithContentsOfURL:request.fileURL options:0 error:&error];
    if (!data) {
        handler(nil, error);
        return;
    }
    Gear *gear = [[Gear alloc] init];
    if (![gear loadFrom:data error:&error]) {
        handler(nil, error);
        return;
    }
    
     handler([QLThumbnailReply replyWithContextSize:request.maximumSize drawingBlock:^BOOL(CGContextRef  _Nonnull context) {
         [gear drawInContext:context size:request.maximumSize];
         return YES;
     }], nil);
}

@end

File: ThumbnailExtension/ThumbnailProvider.swift
//
//  ThumbnailProvider.swift
//  ThumbnailExtension
//
//  Created by Torsten Kammer on 12.11.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Cocoa
import QuickLookThumbnailing

@objc class ThumbnailProvider: QLThumbnailProvider {
    
    override func provideThumbnail(for request: QLFileThumbnailRequest, _ handler: @escaping (QLThumbnailReply?, Error?) -> Void) {
        do {
            let gear = try Gear.load(fromUrl: request.fileURL)
            
            let reply = QLThumbnailReply(contextSize: request.maximumSize) { context in
                gear.draw(in: context, size: request.maximumSize)
                return true
            }
            handler(reply, nil)
        } catch {
            handler(nil, error)
        }
    }
    
}
